src/sentry/static/sentry/app/views/settings/components/forms/choiceMapper.jsx

```diff
@@ -0,0 +1,188 @@
+    /**
+     * The label to show above the row name selected from the dropdown.
+     */
+    mappedColumnLabel: PropTypes.node,
+    /**
+     * A list of column labels (headers) for the multichoice table. This should
+     * have the same number of items as the mappedSelectors prop.
+     */
+    columnLabels: PropTypes.objectOf(PropTypes.node).isRequired,
+    /**
+     * A list of select field properties that should be used to render the
+     * select field for each column in the row.
+     */
+    mappedSelectors: PropTypes.objectOf(PropTypes.objectOf(SelectControl.propTypes))
+      .isRequired,
+  };
+
+  hasValue = value => defined(value) && !objectIsEmpty(value);
+
+  renderField = props => {
+    const {
+      onChange,
+      onBlur,
+      addButtonText,
+      addDropdown,
+      mappedColumnLabel,
+      columnLabels,
+      mappedSelectors,
+      disabled,
+    } = props;
+
+    const mappedKeys = Object.keys(mappedSelectors);
+    const emptyValue = mappedKeys.reduce((a, v) => ({...a, [v]: null}), {});
+
+    const valueIsEmpty = this.hasValue(props.value);
+    const value = valueIsEmpty ? props.value : {};
+
+    const addRow = data => {
+      onChange({...value, [data.value]: emptyValue}, {});
+      onBlur();
+    };
+
+    const removeRow = itemKey => {
+      const updatedValue = {...value};
+      delete updatedValue[itemKey];
+      onChange(updatedValue, {});
+      onBlur();
+    };
+
+    const setValue = (itemKey, fieldKey, fieldValue) => {
+      onChange({...value, [itemKey]: {...value[itemKey], [fieldKey]: fieldValue}}, {});
+      onBlur();
+    };
+
+    // Remove already added values from the items list
+    const selectableValues = addDropdown.items.filter(
+      i => !value.hasOwnProperty(i.value)
+    );
+
+    const valueMap = addDropdown.items.reduce((map, item) => {
+      map[item.value] = item.label;
+      return map;
+    }, {});
+
+    const dropdown = (
+      <DropdownAutoComplete
+        {...addDropdown}
+        alignMenu={valueIsEmpty ? 'right' : 'left'}
+        items={selectableValues}
+        onSelect={addRow}
+        disabled={disabled}
+      >
+        {({isOpen}) => (
+          <DropdownButton
+            icon="icon-circle-add"
+            isOpen={isOpen}
+            size="xsmall"
+            disabled={disabled}
+          >
+            {addButtonText ? addButtonText : t('Add Item')}
+          </DropdownButton>
+        )}
+      </DropdownAutoComplete>
+    );
+
+    // The field will be set to inline when there is no value set for the
+    // field, just show the dropdown.
+    if (!valueIsEmpty) return <Box>{dropdown}</Box>;
+
+    return (
+      <React.Fragment>
+        <Flex align="center">
+          <LabelColumn>
+            <StyledHeader>{mappedColumnLabel}</StyledHeader>
+          </LabelColumn>
+          {mappedKeys.map((k, i) => {
+            const header = <StyledHeader>{columnLabels[k]}</StyledHeader>;
```

Comment:
imo this would read better if it was all just inline, otherwise you have to jump around while reading it.

```javascript
{mappedKeys.map((k, i) => (
  <Box key={k} ml={1} flex="1 0 0">
     <Flex align="center" justify={'space-between'}>
       <StyledHeader>{columnLabels[k]}</StyledHeader>
       {i > mappedKeys.length && dropdown}
     </Flex>
  </Box>
))}

Reply:
e100da0 agreed